torch/csrc/distributed/rpc/utils.cpp

```diff
@@ -101,6 +103,179 @@ std::unique_ptr<RpcCommandBase> deserializeResponse(const Message& response) {
+  }
+  ++ptr; // past the final \n
+
+  std::unordered_map<std::string, std::pair<const char*, size_t>> out;
+  for (const auto& header_ent : header_ents) {
+    out[header_ent.first] = {ptr, header_ent.second};
+    ptr += header_ent.second;
+  }
+  if (!ok || ptr != endp) {
+    throw std::runtime_error("failed bounds");
+  }
+  return out;
+}
+
+static const char* kMeta = "meta";
+static const char* kPayload = "payload";
+}; // namespace
+
+std::string wireSerialize(
+    const std::vector<char>& payload,
+    const std::vector<at::Tensor>& tensors) {
+  struct Ent {
+    std::string name;
+    const char* data;
+    size_t size;
+  };
+  std::vector<Ent> entries;
+  std::string metaEntry;
+  if (!payload.empty()) {
+    entries.push_back({kPayload, payload.data(), payload.size()});
+  }
+
+  if (!tensors.empty()) {
+    torch::jit::Pickler pickler(
+        [&](const void* buf, size_t sz) -> size_t {
+          metaEntry.append(static_cast<const char*>(buf), sz);
+          return sz;
+        },
+        nullptr);
+    pickler.protocol();
+    pickler.pushIValue(tensors);
+    pickler.stop();
+    auto writeable_tensors = pickler.tensorData();
+    entries.push_back({kMeta, metaEntry.data(), metaEntry.size()});
+    for (size_t i = 0; i < writeable_tensors.size(); i++) {
+      entries.push_back({c10::to_string(i),
+                         writeable_tensors[i].data(),
+                         writeable_tensors[i].sizeInBytes()});
+    }
+  }
+
+  std::string header;
+  size_t tot = 0;
+  for (const auto& e : entries) {
+    tot += e.size;
+    header.append(e.name)
+        .append(" ")
+        .append(c10::to_string(e.size))
+        .append("\n");
+  }
+  header.push_back('\n');
+
+  std::string out;
+  out.reserve(header.size() + tot);
+  out.append(header);
+  for (const auto& e : entries) {
+    out.append(e.data, e.size);
+  }
+  return out;
+}
+
+std::pair<std::vector<char>, std::vector<at::Tensor>> wireDeserialize(
+    const void* data,
+    size_t data_size) {
+  auto sections = parseWireSections(data, data_size);
+
+  std::vector<char> payload;
+  auto payload_it = sections.find(kPayload);
+  if (payload_it != sections.end() && payload_it->second.second != 0) {
+    payload.assign(
+        payload_it->second.first,
+        payload_it->second.first + payload_it->second.second);
+  }
+
+  std::vector<at::Tensor> tensors;
+  auto meta_it = sections.find(kMeta);
+  if (meta_it != sections.end()) {
+    const auto& meta_data = meta_it->second;
+    size_t meta_data_pos = 0;
+    auto meta_data_read_func = [&](char* buf, size_t n) -> size_t {
+      if (meta_data_pos >= meta_data.second || n == 0) {
+        return 0;
+      }
+      size_t to_copy =
+          std::min(meta_data_pos + n, meta_data.second) - meta_data_pos;
+      memcpy(buf, meta_data.first + meta_data_pos, to_copy);
+      meta_data_pos += to_copy;
+      return to_copy;
+    };
+    auto section_read_func = [&](const std::string& fname) -> at::DataPtr {
```

Comment:
What does this do? Why it needs a file name? Or is it actually Ent name?

Reply:
It really is an entity name, I can change to that.

fwiw, in the context of torch::save() it kind of is a filename, since torch::save uses zip container format, with a directory full of these entities. But "entity" is probably more clear.